docs/filtering.md

```diff
@@ -0,0 +1,186 @@
+
+- packet-based filtering which filters packets based on their content
+  (headers).
+- metadata-based filtering which filters packets based on their
+  associated metadata.
+
+These filtering mechanisms ensure that only relevant packets are
+reported, so reducing the volume of uninteresting events and
+improving the efficiency of packet tracing.
+
+## Packet
+
+Packet filtering uses a pcap-filter syntax. See `man pcap-filter` for an
+overview on the syntax. Upon execution the pcap-filter gets compiled in cBPF
+and subsequently translated into an eBPF program which in turn gets consumed by
+the probes.
+
+```none
+$ retis collect -f 'tcp port 443'
+...
+```
+
+Packet filtering can be of two types: L2 and L3. Retis automatically detects and
+generates L2/L3 filters based on the expression. This allows to match both packets
+fully formed and packets not having a valid L2 header yet (`sk_buff` having invalid
+`mac_header` but valid and set `network_header`). One advantage of this approach is
+the ability to match locally generated packets while still allowing matches based
+on L2 criteria.
+
+For example, the following filter:
+
+```none
+$ retis collect -f 'arp or tcp port 443'
+L2+L3 packet filter(s) loaded
+...
+```
+
+Internally generates two filters. For probes where only the `network_header`
+is valid and set in the `sk_buff` the filter would match packets with tcp
+source or destination port 443. For `sk_buff` with valid `mac_header` both arp
+and tcp packets would be matched.
+Please note that some limitations exist and they are a direct consequence of
+libpcap capabilities.
+For example filters like:
+
+```none
+$ retis collect -f 'ether broadcast or tcp port 443'
+L2 packet filter(s) loaded
+...
+```
+
+Will only generate L2 filters, that is, packets will be matched only if
+`mac_header` is set.
+For further information about the reason an L3 filter gets skipped, please
+use the `--log-level debug` option, i.e.:
+
+```none
+$ retis --log-level debug collect -f 'ether broadcast or tcp port 443'
+...
+DEBUG Skipping L3 filter generation (Could not compile the filter: libpcap error: not a broadcast link).
+INFO  L2 packet filter(s) loaded
+...
+```
+
+## Metadata
+
+Metadata filtering instead allows to write filters that match packets based
+on their metadata.
+These filters can match against any subfield of the `sk_buff` and subsequent
+inner data structures.
+Meta filtering also automatically follows struct pointers, so indirect access to
+structures pointed by an `sk_buff` field is possible.
+A filter expression is represented by the pseudo EBNF grammar below:
+
+```none
+EXPR ::= LHS ' ' OP_RHS | LHS
+OP_RHS ::= OP ' ' NUMBER | EQ_NE ' ' STRING
+LHS ::= 'sk_buff' MEMBER
+MEMBER ::= NEXTIDENT MEMBER | NEXTIDENT
+NEXTIDENT ::= '.' IDENT (':' MASK (':' IDENT)?)?
+IDENT ::= #'[a-zA-Z_][a-zA-Z0-9_]*'
+OP ::= EQ_NE | '<' | '<=' | '>' | '>='
+EQ_NE ::= '==' | '!='
+MASK ::= ('~')? MASK_NUM
+MASK_NUM ::= NUMBER | BIN
```

Comment:
Isn't BIN a number (HEX is)?

Reply:
yes :)
the problem is that NUMBER is used for right-hand side definition and binary nums are not supported.
Do you prefer to s/NUMBER/DEC_HEX/ ?

Reply:
Maybe we could define an helper to parse such numbers and support all variants for all usages?

Reply:
problem is that on RHS negative decimals are allowed.
I guess this would require a little bit more than simply adding a shared helper.
Not a huge deal, but maybe can be done as a follow-up. 
Does this sound ok for you?

While we speak (negative dec), I realize that the doc requires an update 
to match the syntax:

```diff
diff --git a/docs/filtering.md b/docs/filtering.md
index 59522fd9..fbd902f0 100644
--- a/docs/filtering.md
+++ b/docs/filtering.md
@@ -78,7 +78,7 @@ A filter expression is represented by the pseudo EBNF grammar below:
 
 ```none
 EXPR ::= LHS ' ' OP_RHS | LHS
-OP_RHS ::= OP ' ' NUMBER | EQ_NE ' ' STRING
+OP_RHS ::= OP ' ' RHS_NUM | EQ_NE ' ' RHS_STR
 LHS ::= 'sk_buff' MEMBER
 MEMBER ::= NEXTIDENT MEMBER | NEXTIDENT
 NEXTIDENT ::= '.' IDENT (':' MASK (':' IDENT)?)?
@@ -86,10 +86,10 @@ IDENT ::= #'[a-zA-Z_][a-zA-Z0-9_]*'
 OP ::= EQ_NE | '<' | '<=' | '>' | '>='
 EQ_NE ::= '==' | '!='
 MASK ::= ('~')? MASK_NUM
-MASK_NUM ::= NUMBER | BIN
-STRING ::= '"' ASCII '"' | '\'' ASCII '\''
+MASK_NUM ::= BIN | DEC | HEX
+RHS_STR ::= '"' ASCII '"' | '\'' ASCII '\''
 ASCII ::= #'[:ascii:]*'
-NUMBER ::= HEX | DEC
+RHS_NUM ::= HEX | ('-')? DEC
 HEX ::= #'0x[a-fA-F0-9]+'
 DEC ::= #'[0-9]+'
 BIN ::= #'0b[0-1]+'
```

Reply:
> problem is that on RHS negative decimals are allowed. I guess this would require a little bit more than simply adding a shared helper. Not a huge deal, but maybe can be done as a follow-up. Does this sound ok for you?

IMO it's important for the UX to always unify behaviors as much as possible (here allowing the same set of bases for numbers), it should be possible to disallow unsigned numbers after calling that helpers or using a flag as a parameter.

Having said that I don't want to block this only for this reason, if a follow-up is coming.